### PR TITLE
Refactor provider configuration and versioning

### LINKED ISSUE #1024: Firebird: should use true/false literals for boolean fields instead of 1/0
linq2db generates following SQL

``` sql
-- Firebird
SELECT
	t1.ID,
	t1.Name,
	t1.IsDeleted
FROM
	Product t1
WHERE
	1 <> t1.IsDeleted
```
and it results to exception:
```
FirebirdSql.Data.FirebirdClient.FbException
  HResult=0x80004005
  Message=conversion error from string "1"
  Source=FirebirdSql.Data.FirebirdClient
  StackTrace:
   at FirebirdSql.Data.FirebirdClient.FbCommand.Fetch() in C:\Users\Jiri\Documents\devel\NETProvider\working\Provider\src\FirebirdSql.Data.FirebirdClient\FirebirdClient\FbCommand.cs:line 651
   at FirebirdSql.Data.FirebirdClient.FbDataReader.Read() in C:\Users\Jiri\Documents\devel\NETProvider\working\Provider\src\FirebirdSql.Data.FirebirdClient\FirebirdClient\FbDataReader.cs:line 207
   at LinqToDB.Linq.QueryRunner.<ExecuteQuery>d__2b`1.MoveNext() in c:\projects\linq2db\Source\Linq\QueryRunner.cs:line 329
   at System.Collections.Generic.List`1..ctor(IEnumerable`1 collection)
   at System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source)
   at ConsoleApp1.Program.Main(String[] args) in c:\users\bambr\documents\visual studio 2017\Projects\ConsoleApp1\ConsoleApp1\Program.cs:line 27
```

### DDL
``` sql
CREATE TABLE PRODUCT (
    ID         INTEGER NOT NULL,
    NAME       VARCHAR(64) NOT NULL,
    ISDELETED  BOOLEAN NOT NULL
);
```

### Code example

```c#
using System.Linq;
using LinqToDB.Data;
using LinqToDB.Mapping;

namespace ConsoleApp1
{
    [Table]
    public class Product
    {
        [PrimaryKey]
        public int ID { get; set; }

        [Column(CanBeNull = false, Length = 64)]
        public string Name { get; set; }

        [Column(CanBeNull = false)]
        public bool IsDeleted { get; set; }
    }

    class Program
    {
        static void Main(string[] args)
        {
            using (var conn = new DataConnection())
            {
                var products = from p in conn.GetTable<Product>() where !p.IsDeleted select p;
                var list = products.ToList();
            }
        }
    }
}

```
Valid SQL should be
``` sql
-- Firebird
SELECT
	t1.ID,
	t1.Name,
	t1.IsDeleted
FROM
	Product t1
WHERE
	true <> t1.IsDeleted
```

### Environment details
linq2db version: *1.10.0*  
Database Server: *Firebird 3.0.3*  
Database Provider: *FirebirdSql.Data.FirebirdClient 5.12.1*
Operating system: *Windows 10*  
Framework version: *.NET Framework 4.6.2*

### LINKED ISSUE #3154: CROSS and OUTER APPLY when using hana provider
Hello !
Nested many to many associations end up generating this query but APPLY is not supported by HANA

```sql
     DECLARE @p_1  -- Int32
      SET     @p_1 = 50
      
      SELECT
        "key_data_result"."KUNNR",
        "key_data_result"."MANDT",
        "t2"."Id",
        "t2"."Name",
        "t2"."BUKRS",
        "t2"."MANDT",
        "t2"."Name_1"
      FROM
        (
                SELECT DISTINCT
                        "t1"."KUNNR",
                        "t1"."MANDT"
                FROM
                        (
                                SELECT
                                        "p"."KUNNR",
                                        "p"."MANDT"
                                FROM
                                        "SAPABAP1"."KNA1" "p"
                                WHERE
                                        "p"."MANDT" = '100' AND Lower("p"."SORTL") = 'webportal'
                                ORDER BY
                                        "p"."ERDAT" DESC
                                LIMIT :"p_1"
                        ) "t1"
        ) "key_data_result"
                CROSS APPLY (
                        SELECT
                                "detail"."WERKS" as "Id",
                                "detail"."NAME1" as "Name",
                                "c_1"."BUKRS",
                                "c_1"."MANDT",
                                "c_1"."BUTXT" as "Name_1"
                        FROM
                                "SAPABAP1"."T001W" "detail"
                                        LEFT JOIN "SAPABAP1"."T001K" "a_CompanyAgencyAssociation" ON "detail"."WERKS" = "a_CompanyAgencyAssociation"."BWKEY" AND "detail"."MANDT" = "a_CompanyAgencyAssociation"."MANDT"
                                        LEFT JOIN "SAPABAP1"."T001" "c_1" ON "c_1"."BUKRS" = "a_CompanyAgencyAssociation"."BUKRS" AND "c_1"."MANDT" = "a_CompanyAgencyAssociation"."MANDT"
                        WHERE
                                EXISTS(
                                        SELECT
                                                *
                                        FROM
                                                "SAPABAP1"."KNVV" "e"
                                        WHERE
                                                "key_data_result"."KUNNR" = "e"."KUNNR" AND
                                                "key_data_result"."MANDT" = "e"."MANDT" AND
                                                "e"."VKORG" = "detail"."WERKS" AND
                                                "e"."MANDT" = "detail"."MANDT"
                                )
                ) "t2"
```
```c#
        // class Customer
        [GraphQLIgnore]
        [Association(ThisKey = "Id, Mandt", OtherKey = "CustomerId, Mandt")]
        public CustomerAgencyAssociation[] CustomerAgencyAssociation { get; set; }

        [Association(QueryExpressionMethod = nameof(AgencyExpression), CanBeNull = true)]
        public Agency[] Agencies { get; set; }

        public static Expression<Func<Customer, IDataContext, IQueryable<Agency>>> AgencyExpression =>
        (c, db) => db.GetTable<Agency>().Where(a => c.CustomerAgencyAssociation.Any(e => e.AgencyId == a.Id && e.Mandt == a.Mandt));
```
```c#
        // class Agency

        [GraphQLIgnore]
        [Association(ThisKey = "Id, Mandt", OtherKey = "AgencyId, Mandt")]
        public CompanyAgencyAssociation CompanyAgencyAssociation { get; set; }

        [Association(QueryExpressionMethod = nameof(AgencyExpression), CanBeNull = true)]
        public Agency[] Agencies { get; set; }

        public static Expression<Func<Agency, IDataContext, IQueryable<Company>>> CompanyExpression =>
        (r, db) => db.GetTable<Company>().Where(c => c.Id == r.CompanyAgencyAssociation.CompanyId && c.Mandt == r.CompanyAgencyAssociation.Mandt);

```
```c#
            var test = _connection.Customers.LoadWith(e => e.Agencies).ThenLoad(e => e.Company).ToList();
```

image

### PR DESCRIPTION
**Pull request by MaceWindu**
Fixes #1024
Fixes #3154

This PR reviews database providing versioning implementation:

 - updates API to register database provider in options to use same set of methods for each provider. Old implementation had non-uniform set of registration methods for each database, often even with different names for same database provider. In new implementation each provider will have 4 overloads to register provider (except SQL CE provider, which doesn't have two last methods as it doesn't have versioning):
   - `Use<DB>(optionSetter)`
   - `Use<DB>(connectionString, optionSetter)`
   - `Use<DB>(dialect and/or provider, optionSetter = null)`
   - `Use<DB>(dialect and/or provider, connectionString, optionSetter = )`

 - major cleanup of `<DB>Tools` classes and related code:
   - removed obsoleted `BulkCopy` APIs
   - removed obsoleted settings
   - moved provider detection logic to corresponding `ProviderDetector` classes where it wasn't done befor
   - exposed `AutoDetect` option property for provider detector
   - provider factories (`IDataProviderFactory`) were updated to support new versioning options like dialect and provider assembly (`assemblyName`, `version` attributes). (I wonder if anyone uses those factories at all)
   - unify `GetDataProvider` API to accept provider and dialect (where applicable) and connection string parameters
   - unify `CreateDataConnection` API to be 3 methods with
     - connection string/connection/transaction as first parameter
     - provider and dialect as other parameters where applicable

 - added dialect/provider versioning to providers missing it:
   - all provider and dialect enums have `AutoDetect` value as first entry to use provider/dialect detection logic (still we recommend to specify them explicitly to not depend on it if you don't need to work in unknown environment)
   - `Access`: add `AccessProvider` provider selection enum
   - `Firebird`: add `FirebirdVersion` dialect selection enum
   - `Informix`: add `InformixProvider` provider selection enum
   - `MySql`/`MariaDB`: add `MySqlProvider` provider and `MySqlVersion` dialect selection enums
   - `SQLite`: add `SQLiteProvider` provider selection enum
   - `SAP HANA`: add `SapHanaProvider` provider selection enum
   - `SAP/Sybase ASE`: add `SybaseProvider` provider selection enum

 - database connection creation logic moved from database provider to database provider adapter interface (`IDynamicProviderAdapter`)
 - fixed issue with database dialect detection when options specify connection or transaction but not connection string
 - simplify struct types registration in mapping schema: now nullable type will be registered automatically when you register struct type, so there is no need to define two same mappings anymore. Apllicable to `AddScalarType` and `SetDataType` methods.

 - adds dialects support for `Firebird`. Instead of old implementation with generated SQL compatible with Firebird 2.5 now we support 4 dialects: 2.5, 3.0, 4.0 and 5.0
   - `v3+`: `bool` now maps to `BOOLEAN` type by default. To use old mapping to `1`/`0` chars, specify `DataType.Char` type hint for your mappings or use 2.5 dialect
   - `v3+`: switched to `OFFSET/FETCH` clauses
   - `v4+`: enabled `LATERAL` joins
   - `v3+`: increased batch size for bulk copy queries from 64Kb as query size to 10Mb
   - `v5+`: increased supported elements limit in `IN` predicate from 1 500 to 65 535 items
   - `v5+`: enable support for `WHEN NOT MATCHED BY SOURCE` merge operations
   - `v5+`: enable support for `quarter` qualifier for `DatePart` functions
   - more dialect-specific changes will be added later by separate PR

 - adds dialects support for `MySQL/MariaDB`. Instead of old implementation compatible with MySQL 5.5, now we have support for 4 dialects: MySQL 5.7, MySQL 8.0 and MariaDB 10
   - `MySQL 8.0`/`MariaDB`: don't use `FROM DUAL` for table-less queries anymore (still needed for `MySQL 5.7` table-less queries with filter)
   - `MySQL 8.0`/`MariaDB`: enabled support for `CAST` to `FLOAT`/`DOUBLE` types
   - `MySQL 8.0`: enabled `LATERAL` joins
   - `MySQL 8.0`/`MariaDB`: enabled native support for `EXCEPT/INTERSECT ALL/DISTINCT` set operators
   - `All`: remove generation of `FLOAT(N)` type. Use `DbType` property in your mappings if you still need it for some reason
   - more dialect-specific changes will be added later by separate PR

 - enabled `LATERAL` joins support for SAP HANA

For developers:

 - Firebird 5 added to testing
- you will need to update your test configuration configs to use new test provider names for MySQL and Firebird:
  - `MySql.5.7`: MySQL 5.7 using `MySql.Data` provider
  - `MySqlConnector.5.7`: MySQL 5.7 using `MySqlConnector` provider
  - `MySql.8.0`: MySQL 8.0 using `MySql.Data` provider
  - `MySqlConnector.8.0`: MySQL 8.0 using `MySqlConnector` provider
  - `MariaDB.11`: MariaDB 11 using `MySqlConnector` provider. Note that currently MariaDB 11 doesn't differ from 10 in SQL features, so we don't add separate MariaDB 11 dialect.
 - `Firebird.2.5`
 - `Firebird.3`
 - `Firebird.4`
 - `Firebird.5`
- we don't test MySQL 5.5 anymore and bumped lowest supported version to 5.7. This also helped to discover some tests incorrectly labeled as 5.7+ compatible while being working only with MySQL 8.0
- `year(2)` pre-mysql 5.7 legacy type removed from testing

Known issues (introduced or discovered by PR):
 - Firebird 2.5 `bool` to `char` mapping still configured in `FirebirdProvider.SetParameter` method instead of `MappingSchema` due to a lot discovered bugs with `DataParameter`-based mappings. Will be re-enabled and fixed in new parser branch
 - SAP HANA lateral sub-queries with parameters doesn't work as we need to convert parameters to constants (SAP HANA limitation). Will be done in new parser branch
 - Some SAP HANA queries are broken after `LATERAL` joins were enabled (bad optimizer?). Should be already fixed in new parser branch
 - `EXCEPT/INTERSECT ALL` set operators emulation for unsupporting databases doesn't handle duplicate records. Will be fixed in new parser/separate PR
 - Some firebird tests generate `CROSS APPLY` for Firebird 5 after `LATERAL` joins enabled. Should be already fixed in new parser branch

### CODE REVIEW

File: Source/LinqToDB/DataProvider/Access/AccessProvider.cs [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+		/// </summary>
+		AutoDetect,
+		/// <summary>
+		/// Access OLE DB provider.
+		/// </summary>
+		OleDb,
+		/// <summary>
+		/// Access ODBC provider.
+		/// </summary>
+		ODBC
```
Comment (viceroypenguin) [inline Source/LinqToDB/DataProvider/Access/AccessProvider.cs:19]:
```suggestion
		ODBC,
```

File: Source/LinqToDB/DataProvider/Access/AccessProviderDetector.cs [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+		public AccessProviderDetector() : base(default, default)
+		{
+		}
+
+		static readonly Lazy<IDataProvider> _accessOleDbDataProvider = DataConnection.CreateDataProvider<AccessOleDbDataProvider>();
+		static readonly Lazy<IDataProvider> _accessODBCDataProvider  = DataConnection.CreateDataProvider<AccessODBCDataProvider>();
+
+		public override IDataProvider? DetectProvider(ConnectionOptions options)
+		{
+			if (options.ConnectionString?.Contains("Microsoft.ACE.OLEDB") == true || options.ConnectionString?.Contains("Microsoft.Jet.OLEDB") == true)
```
Comment (viceroypenguin) [inline Source/LinqToDB/DataProvider/Access/AccessProviderDetector.cs:23]:
```suggestion
			if (options.ConnectionString?.Contains("Microsoft.ACE.OLEDB") == true 
			    || options.ConnectionString?.Contains("Microsoft.Jet.OLEDB") == true)
```

File: Source/LinqToDB/DataProvider/Access/AccessProviderDetector.cs [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+				return _accessOleDbDataProvider.Value;
+			}
+
+			if (options.ProviderName == ProviderName.AccessOdbc
+				|| options.ConfigurationString?.Contains("Access.Odbc") == true)
+			{
+				return _accessODBCDataProvider.Value;
+			}
+
+			if (options.ProviderName == ProviderName.Access || (options.ConfigurationString?.Contains("Access") == true && !options.ConfigurationString.Contains("DataAccess")))
```
Comment (viceroypenguin) [inline Source/LinqToDB/DataProvider/Access/AccessProviderDetector.cs:34]:
```suggestion
			if (options.ProviderName == ProviderName.Access
			    || (options.ConfigurationString?.Contains("Access") == true
			        && !options.ConfigurationString.Contains("DataAccess")))
```

File: Source/LinqToDB/DataProvider/Access/AccessProviderDetector.cs [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+		public override Dialect? DetectServerVersion(DbConnection connection)
+		{
+			return default(Dialect);
+		}
+
+		protected override DbConnection CreateConnection(AccessProvider provider, string connectionString)
+		{
+			return provider == AccessProvider.ODBC
+				? OdbcProviderAdapter.GetInstance().CreateConnection(connectionString)
+				: OleDbProviderAdapter.GetInstance().CreateConnection(connectionString);
```
Comment (viceroypenguin) [inline Source/LinqToDB/DataProvider/Access/AccessProviderDetector.cs:83]:
```suggestion
			var adapter = provider == AccessProvider.ODBC
				? OdbcProviderAdapter
				: OleDbProviderAdapter;
			return adapter.GetInstance().CreateConnection(connectionString)
```

**Review by viceroypenguin — APPROVED** [review]

**Review by sdanyliv — DISMISSED** [review]

**Review by sdanyliv — APPROVED** [review]

**Comment by MaceWindu** [comment]
PR summary added

**Comment by MaceWindu** [comment]
baselines good, except one sqlite test and missing FB features (all fixed) and several bad queries generated for sap hana after lateral enabled (will be fixed in parser branch, but we really need to add asserts to those tests )

**Review by sdanyliv — APPROVED** [review]

**Review by viceroypenguin — APPROVED** [review]

**Review by sdanyliv — APPROVED** [review]

**Pull request merged**